src/CP/core/search/ilds.jl

```diff
@@ -0,0 +1,77 @@
+
+This function fills the toCall Stack (LIFO) and perform a recursive Limited Discrepancy Search. Some procedures will contain a call to `expandIlds!` itself. Each `expandIlds!` 
+call is wrapped around a `saveState!` and a `restoreState!` to be able to backtrack thanks to the trailer. Depth is the virtual depth of the "bounding tree", and decrease as 
+long as variables get bounded. Discrepancy is the virtual Discrepancy of the path in the "bounding tree". It is not updated each time the algorithm steps into the right sub-tree, but after
+each right step that bounded one or more variable. ( after the fix-point algorithm ) This method allows to work with search tree where the depth ( and hence the max discrepancy ) 
+is unknown.
+    
+This implementation is based on this paper : Improved Limited Discrepancy Search - 1996 - Richard E. Korf
+"""
+function expandIlds!(toCall::Stack{Function}, discrepancy::Int64, previousdepth::Int64, direction::Union{Nothing, Symbol} , model::CPModel, variableHeuristic::AbstractVariableSelection, valueSelection::ValueSelection, newConstraints=nothing)
+    # Dealing with limits
+    model.statistics.numberOfNodes += 1
+    if !belowNodeLimit(model)
+        return :NodeLimitStop
+    end
+    if !belowSolutionLimit(model)
+        return :SolutionLimitStop
+    end
+
+    # Fix-point algorithm
+    feasible, pruned = fixPoint!(model, newConstraints)
+    if !feasible
+        return :Infeasible
+    end
+    if solutionFound(model)
+        if (direction ==:Left && discrepancy == 0)  || (direction ==:Right && discrepancy == 1 ) || isnothing(direction)
+            triggerFoundSolution!(model) 
+
+        end
+        return :FoundSolution 
+    end
+    depth = sum([!isbound(v) for (k,v) in model.variables])  #recomputed at each step, maybe not efficient
+    discrepancy =(depth != previousdepth && direction == :Right ) ? discrepancy - 1 : discrepancy   
+    @assert depth > 0 "a least one variable should be branchable"
+    # Variable selection
+    x = variableHeuristic(model)
+
+    # Value selection
+    v = valueSelection(DecisionPhase(), model, x, nothing)
+
+    if (discrepancy>=0)   
+        push!(toCall, (model) -> (restoreState!(model.trailer); :BackTracking))
+        push!(toCall, (model) -> (remove!(x.domain, v); expandIlds!(toCall,discrepancy, depth, :Right, model, variableHeuristic, valueSelection, getOnDomainChange(x))))
+        push!(toCall, (model) -> (saveState!(model.trailer); :SavingState))
+    end
+
+    if (depth>discrepancy)
+        push!(toCall, (model) -> (restoreState!(model.trailer); :BackTracking))
+        push!(toCall, (model) -> (assign!(x, v); expandIlds!(toCall,discrepancy, depth-1, :Left, model, variableHeuristic, valueSelection, getOnDomainChange(x))))
+        push!(toCall, (model) -> (saveState!(model.trailer); :SavingState))
+    end
```

Comment:
It would be better if you use the updated version of this code in `expandDfs!`. Otherwise `prunedDomains` won't be set correctly in the constraint propagation, and it would harm `tableCT`.

Reply:
Okey thanks for the warning Pierre !

Reply:
I just fixed it !